tests/TestAsset/NoUselessInlineAnnotationsRule/fixture.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace SlamPhpStan\Tests\TestAsset\NoUselessInlineAnnotationsRule;
+
+class Foo {
+    public function tester()
+    {
+        /** @var Bar[] $foo1 */
+        $foo1 = $this->bar1();
+
+        /** @var Bar $foo2 */
+        $foo2 = $this->bar2();
```

Comment:
@ondrejmirtes after using more and more @phpstan I now trust it much more than developers annotations: now I'm scared that in the future this kind of redundant annotation (`::bar2` already returns `Bar`) will hide bugs if the related API change.

Is such a rule, `NoRedundantInlineAnnotationsRule`, feasable?

Reply:
Hi, first of all, make sure to run PHPStan with `conf/bleedingEdge.neon` (include it in phpstan.neon), it contains some additional rules around `@var`.

But the main problem is that you need to consider these things:

1) `@var` is sometimes used for more precise casting - I get object and I know it's `Foo`.
2) `@var` is sometimes used for fixing wrong 3rd party phpDoc - I get `string` and I know it's `int`. This point makes any type verification impossible.
3) Sometimes `@var` is used because PHPStan understands the type, but IDE doesn't.

These points make any general rule impossible. But some more proprietary variant is possible to implement, very easily. I'd recommend you to look how those bleeding edge rules are implemented in 0.11.9 (https://github.com/phpstan/phpstan/releases/tag/0.11.9) - commits are linked in the release notes.